crates/core/src/doc/field.rs

```diff
@@ -146,261 +146,395 @@ impl Document {
-					match perms {
-						// The field PERMISSIONS clause
-						// is FULL, enabling this field
-						// to be updated without checks.
-						Permission::Full => (),
-						// The field PERMISSIONS clause
-						// is NONE, meaning that this
-						// change will be reverted.
-						Permission::None => val = old.as_ref().clone(),
-						// The field PERMISSIONS clause
-						// is a custom expression, so
-						// we check the expression and
-						// revert the field if denied.
-						Permission::Specific(e) => {
-							// Arc the current value
-							let now = Arc::new(val.clone());
-							// Disable permissions
-							let opt = &opt.new_with_perms(false);
-							// Configure the context
-							let mut ctx = MutableContext::new(ctx);
-							ctx.add_value("before", old.clone());
-							ctx.add_value("input", inp.clone());
+				// The outer value should be a record
+				else {
+					// There was a field check error
+					return Err(Error::FieldCheck {
+						thing: self.rid.to_string(),
+						field: self.def.name.clone(),
+						check: kind.to_string(),
+						value: val.to_string(),
+					});
+				}
+			}
+			// This is not the `id` field
+			else {
+				// Check the type of the field value
+				let val = val.coerce_to(kind).map_err(|e| match e {
+					// There was a conversion error
+					Error::CoerceTo {
+						from,
+						..
+					} => Error::FieldCheck {
+						thing: self.rid.to_string(),
+						field: self.def.name.clone(),
+						check: kind.to_string(),
+						value: from.to_string(),
+					},
+					// There was a different error
+					e => e,
+				})?;
+				// Return the modified value
+				return Ok(val);
+			}
+		}
+		// Return the original value
+		Ok(val)
+	}
+	/// Process any DEFAULT clause for the field definition
+	async fn process_default_clause(&mut self, val: Value) -> Result<Value, Error> {
+		// This field has a value specified
+		if !val.is_none() {
+			return Ok(val);
+		}
+		// The document is not being created
+		if !self.doc.is_new() {
+			return Ok(val);
+		}
+		// Get the default value
+		let def = match &self.def.default {
+			Some(v) => Some(v),
+			_ => match &self.def.value {
+				// The VALUE clause doesn't
+				Some(v) if v.is_static() => Some(v),
+				_ => None,
+			},
+		};
+		// Check for a DEFAULT clause
+		if let Some(expr) = def {
+			// Arc the current value
+			let now = Arc::new(val);
+			// Get the current document
+			let doc = Some(&self.doc.current);
+			// Configure the context
+			let ctx = match self.context.take() {
+				Some(mut ctx) => {
+					ctx.add_value("after", now.clone());
+					ctx.add_value("value", now);
+					ctx
+				}
+				None => {
+					let mut ctx = MutableContext::new(self.ctx);
+					ctx.add_value("before", self.old.clone());
+					ctx.add_value("input", self.inp.clone());
+					ctx.add_value("after", now.clone());
+					ctx.add_value("value", now);
+					ctx
+				}
+			};
+			// Freeze the new context
+			let ctx = ctx.freeze();
```

Comment:
would it be worth generalizing this inside a function across the various methods?